src/services/helpers/workspaceAdapter.ts

```diff
@@ -0,0 +1,229 @@
+      return (status || '') as keyof typeof WorkspaceStatus;
+    }
+  }
+
+  set status(status: keyof typeof WorkspaceStatus) {
+    if (isWorkspaceV1(this.workspace)) {
+      this.workspace.status = status;
+    } else {
+      (this.workspace as IDevWorkspace).status.phase = status;
+    }
+  }
+
+  get ideUrl(): string | undefined {
+    if (isWorkspaceV1(this.workspace)) {
+      const runtime = this.workspace.runtime;
+      if (!runtime || !runtime.machines) {
+        return;
+      }
+      for (const machineName of Object.keys(runtime.machines)) {
+        const servers = runtime.machines[machineName].servers || {};
+        for (const serverId of Object.keys(servers)) {
+          const attributes = (servers[serverId] as any).attributes;
+          if (attributes && attributes['type'] === 'ide') {
+            return servers[serverId].url;
+          }
+        }
+      }
+    } else {
+      return (this.workspace as IDevWorkspace).status.ideUrl;
+    }
+  }
+
+  get storageType(): che.WorkspaceStorageType {
+    if (isWorkspaceV1(this.workspace)) {
+      return attributesToType(this.workspace.devfile.attributes);
+    } else {
+      throw new Error('not implemented.');
+    }
+  }
+
+  set storageType(type: che.WorkspaceStorageType) {
+    if (isWorkspaceV1(this.workspace)) {
+      const attributes = typeToAttributes(type);
+      Object.assign(this.workspace.devfile.attributes, attributes);
+    } else {
+      throw new Error('not implemented.');
+    }
+  }
+
+  get devfile(): che.WorkspaceDevfile | IDevWorkspaceDevfile {
+    if (isWorkspaceV1(this.workspace)) {
+      return this.workspace.devfile as T extends che.Workspace ? che.WorkspaceDevfile : IDevWorkspaceDevfile;
+    } else {
+      return devWorkspaceToDevfile((this.workspace as IDevWorkspace)) as T extends che.Workspace ? che.WorkspaceDevfile : IDevWorkspaceDevfile;
+    }
+  }
+
+  set devfile(devfile: che.WorkspaceDevfile | IDevWorkspaceDevfile) {
+    if (isWorkspaceV1(this.workspace)) {
+      this.workspace.devfile = devfile as che.WorkspaceDevfile;
+    } else {
+      (this.workspace as IDevWorkspace) = devfileToDevWorkspace(devfile as IDevWorkspaceDevfile, this.status === WorkspaceStatus[WorkspaceStatus.RUNNING]);
+    }
+  }
+
+  get projects(): string[] {
+    if (isWorkspaceV1(this.workspace)) {
+      return (this.workspace.devfile.projects || [])
+        .map(project => project.name);
+    } else {
+      const devfile = devWorkspaceToDevfile(this.workspace as IDevWorkspace);
+      return (devfile.projects || [])
+        .map(project => project.name);
+    }
+  }
+
+}
+
+export function convertWorkspace<T extends che.Workspace | IDevWorkspace>(workspace: T): Workspace {
+  return new WorkspaceAdapter(workspace);
+}
+
+export function isWorkspaceV1(workspace: any): workspace is che.Workspace {
+  return !isWorkspaceV2(workspace);
+}
+
+export function isWorkspaceV2(workspace: che.Workspace | IDevWorkspace): workspace is IDevWorkspace {
+  return !!(
```

Comment:
I do not like ```!!``` with ```===```
Could you simplify it?
```
return (workspace as IDevWorkspace).kind === 'DevWorkspace';
```

Reply:
fixed